File: haskell/isogram/Iteration1/src/Isogram.hs
module Isogram (isIsogram) where

import Data.Char
import Data.List
import Data.Map (Map)
import qualified Data.Map as Map

isIsogram :: String -> Bool
isIsogram input = all (== 1) (Map.elems letterCounts)
  where
    cleanedInput = filter isLetter (map toLower input)
    updateCount counts c = Map.insertWith (+) c 1 counts
    letterCounts = foldl updateCount Map.empty cleanedInput

File: fsharp/isogram/Iteration1/Isogram.fs
module Isogram

open System

let isIsogram (phrase: string): bool =
    List.ofSeq phrase
        |> List.filter Char.IsLetter
        |> List.map Char.ToLower
        |> List.countBy id
        |> List.map snd
        |> List.forall (fun count -> count = 1)
